content/5_Plat/Bin_Jump.mdx

```diff
@@ -320,10 +317,166 @@ int main() {
 
 <FocusProblem problem={problems.lca[1]} />
 
-### Tutorial
+### Solution (Company Queries II) 
+
+<Resources>
+	<Resource source="CPH" title="18.3 - LCA Method 1">brief description</Resource>
+	<Resource
+		source="Benq"
+		title="LCAjump"
+		url="https://github.com/bqi343/USACO/blob/master/Implementations/content/graphs%20(12)/Trees%20(10)/LCAjump.h"
+	>
+		alternative implementation
+	</Resource>
+</Resources>
+
+### Solution (Company Queries II) 
+
+To find $\textrm{lca}(a, b)$, we can first lift the lower node of $a$ and $b$ to the same depth as the other. Then, we lift both nodes up decrementally. At the end, the parent of either node is the LCA of the two.
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+CodeSnip{Benq Template}
+	
+int N, Q, T = 1;
+int depth[200005];
+int up[200005][20];
+vi adj[200005];
+
+void dfs(int v) {
+	FOR(i, 1, 20) {
+		up[v][i] = up[up[v][i-1]][i-1];
+	}
+
+	for(int x : adj[v]) {
+		if(x != up[v][0]) {
+			depth[x] = depth[up[x][0] = v]+1;
+			dfs(x);
+		}
+	}
+}
+
+int jump(int x, int d) {
+	F0R(i, 20) {
+		if((d >> i) & 1) 
+			x = up[x][i];
+	}
+	return x; 
+}
+
+int LCA(int a, int b) {
+	if(depth[a] < depth[b]) swap(a, b);
+
+	a = jump(a, depth[a] - depth[b]);
+	if(a == b) return a;
+
+	R0F(i, 20) {
+		int aT = up[a][i], bT = up[b][i];
+		if(aT != bT) a = aT, b = bT;
+	}
+
+	return up[a][0];
+}
+	
+int main() {
+	re(N, Q);
+	
+	rep(N-1) {
+		int b; re(b); --b;
+		adj[b].pb(i+1);
+	}
+
+	up[0][0] = 0, depth[0] = 0;
+	dfs(0);
+
+	rep(Q) {
+		int a, b; re(a, b); --a; --b;
+		ps(LCA(a, b)+1);
+	}
+}
+```
+
+</CPPSection>
+
+</LanguageSection>
+
+### Alternative Solution (Company Queries II)
+
+As mentioned in the [Euler Tour Technique](/gold/tree-euler) module, let $\texttt{st}, \texttt{en}, \texttt{up}$ be the time-in, time-out, and ancestor table for all nodes in the tree.
+
+These can be filled with a DFS traversal. $\texttt{up}$ can be generated because DFS allows the ancestors to be filled first before traversing the current node.
+
+With this information, we can declare node $a$ an ancestor of $b$ if $\texttt{st}[a] \le \texttt{st}[b]$ and $\texttt{en}[a] \ge \texttt{en}[b]$
+
+We know that if $a$ is an ancestor of $b$ or $b$ is an ancestor of $a$, the answer will be $a$ and $b$ respectively. Otherwise, we can loop from the largest jumps to the the direct ancestor of $a$ to find the LCA of $a$ and $b$. If $\texttt{up}[a][i]$ is not an ancestor of $b$, then we can set $a$ to be $\texttt{up}[a][i]$, else, we will decrement $i$ and try to find a lower common ancestor. Afterwards, our answer is not $a$, but the parent of $a$. 
```

Comment:
> We know that if $a$ is an ancestor of $b$ or $b$ is an ancestor of $a$, the answer will be $a$ and $b$ respectively

I think this should say "$a$ or $b$ respectively"

> Otherwise, we can loop from the largest jumps to the the direct ancestor of $a$

not sure what this means